### PR TITLE
No shallow taps

### PR DESCRIPTION
**Pull request by cnnrmnn**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
See #11328.
Removes the --full and --shallow tap command flags. Changes tap command to fully clone taps and convert any existing shallow clones to full clones when updated.

### CODE REVIEW

**Review by MikeMcQuaid — APPROVED** [review]
Makes sense to me, nice work @cnnrmnn! Won't merge just yet in case other maintainers have thoughts.

File: Library/Homebrew/cmd/tap.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/tap.rb:34]: Perhaps we should do this automatically if `ENV["CI"]` is set?

Comment (cnnrmnn) [reply Library/Homebrew/cmd/tap.rb:34]: If you mean fetch the tap as a shallow clone, I agree. I can make this change.

**Comment by Bo98** [comment]
Does this need a deprecation path?

File: Library/Homebrew/cmd/tap.rb
Comment (Rylan12) [reply Library/Homebrew/cmd/tap.rb:34]: The full `git` history is needed for commands like `brew audit`, right? Assuming yes, I don't think we should do this.

If a user _must_ have a shallow clone, they can always manually clone into the correct directory, so I don't think we need to provide the option.

**Comment by Rylan12** [comment]
> Does this need a deprecation path?

Yes I think it does. I wouldn't be surprised if there are scripts tat use `brew tap --shallow`.

File: Library/Homebrew/cmd/tap.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/tap.rb:34]: Fine with me.

File: Library/Homebrew/cmd/tap.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/tap.rb:33]: Given the comments about deprecation path (which I agree with): let's leave these options in place but make/document them as a no-op. Add a commented-out `# odeprecated "`brew tap --full`" for each so we deprecate, disable and remove them on the usual minor release cycle. Shout if any of this was badly explained and you need help!

File: Library/Homebrew/cmd/tap.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/tap.rb:61]:
```suggestion
      if args.full?
        opoo "`brew tap --full` is now a no-op!"
        # odeprecated "`brew tap --full`"
      end
      
      if args.shallow?
        opoo "`brew tap --shallow` is now a no-op!"
        # odeprecated "`brew tap --shallow`"
      end

      tap = Tap.fetch(args.named.first)
```

File: Library/Homebrew/cmd/tap.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/tap.rb:30]:
```suggestion
```

File: Library/Homebrew/cmd/tap.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/tap.rb:34]:
```suggestion
```

File: Library/Homebrew/cmd/tap.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/tap.rb:32]:
```suggestion
             description: "Convert a shallow clone to a full clone without untapping. Taps are only cloned as "\
```

File: Library/Homebrew/cmd/tap.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/tap.rb:33]:
```suggestion
                          "shallow clones if `--shallow` was originally passed."
```

**Review by MikeMcQuaid — COMMENTED** [review]
Nice work, almost there!

**Comment by cnnrmnn** [comment]
> Also, we have have a `replacement` option that can be passed to the `switch` that helps with deprecating/disabling command flags. Although it seems, for some reason, that they can only be used for disabling rather than deprecations... Can you add commented out replacement lines for the `--full` and `--shallow` lines?

By disabling, do you mean removing the command altogether or deprecating it as a no-op?

**Comment by cnnrmnn** [comment]
> Does that help clarify?

Yes, thanks!

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks again @cnnrmnn, you rock!

**Comment by cnnrmnn** [comment]
> Thanks again @cnnrmnn, you rock!

codecov is blocking the merge. Is there any action I need to take to resolve this?

**Comment by Bo98** [comment]
Codecov isn't accurate until all other CI has run.

**Comment by cnnrmnn** [comment]
Two tests failed on MacOS. I poked around the cask uninstall code and don't see how it's at all related to any of my changes. Am I missing something?

**Comment by MikeMcQuaid** [comment]
@cnnrmnn Flaky tests, I'll rerun!

**Comment by cnnrmnn** [comment]
> @cnnrmnn Flaky tests, I'll rerun!

Looks like they failed again. I'm unsure why.

**Comment by Bo98** [comment]
Rebase on master and it should work.

**Comment by Rylan12** [comment]
Thanks for the PR, @cnnrmnn! 🎉

Also, I noticed that you're a Yale student—I'll be joining you there next year!

**Comment by Rylan12** [comment]
Thanks, will do! ❤️🐶

**Pull request merged**
